File: soc-server/internal/services/indicator_service.go
// ==============================================================================
// Package services - Indicator Service (IOCs)
// File: internal/services/indicator_service.go
// Mô tả: Business logic cho Indicator (IOC) management. Thay thế MISP.
// ==============================================================================

package services

import (
	"time"

	"soc-server/internal/models"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// DefaultIndicators returns a few starter IOC samples so the SOC dashboard is populated immediately.
func DefaultIndicators() []models.Indicator {
	return []models.Indicator{
		{
			ID:          uuid.New().String(),
			Type:        models.IOCType("ip-src"),
			Value:       "185.220.101.182",
			Category:    "c2_server",
			Source:      "osint",
			MITRETactic: "Command and Control",
			RiskScore:   88,
			Description: "Known malicious IP used in phishing and command-and-control activity.",
			IsActive:    true,
		},
		{
			ID:          uuid.New().String(),
			Type:        models.IOCType("domain"),
			Value:       "login-secure-update.com",
			Category:    "phishing",
			Source:      "analyst",
			MITRETactic: "Phishing",
			RiskScore:   91,
			Description: "Phishing domain used in credential harvesting campaigns.",
			IsActive:    true,
		},
		{
			ID:          uuid.New().String(),
			Type:        models.IOCType("sha256"),
			Value:       "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08",
			Category:    "malware",
			Source:      "internal",
			MITRETactic: "Execution",
			RiskScore:   94,
			Description: "Known malicious payload hash observed in test environment.",
			IsActive:    true,
		},
	}
}

// IndicatorService - Service xử lý business logic cho IOCs
type IndicatorService struct {
	db *gorm.DB
}

// NewIndicatorService - Khởi tạo IndicatorService
func NewIndicatorService(db *gorm.DB) *IndicatorService {
	return &IndicatorService{db: db}
}

// GetAllIndicators - Lấy danh sách IOCs
func (s *IndicatorService) GetAllIndicators(page, pageSize int, iocType string) ([]models.Indicator, int64, error) {
	var indicators []models.Indicator
	var total int64

	query := s.db.Model(&models.Indicator{})
	if iocType != "" {
		query = query.Where("type = ?", iocType)
	}

	query.Count(&total)

	offset := (page - 1) * pageSize
	err := query.Order("created_at DESC").
		Offset(offset).Limit(pageSize).
		Find(&indicators).Error

	return indicators, total, err
}

// SearchIndicatorsCompat performs the value/type lookup used by the legacy
// MISP-shaped n8n node, but searches the SOC App indicators table.
func (s *IndicatorService) SearchIndicatorsCompat(iocType, value string, limit int) ([]models.Indicator, error) {
	if limit <= 0 {
		limit = 50
	}
	if limit > 1000 {
		limit = 1000
	}

	query := s.db.Model(&models.Indicator{}).Where("is_active = ?", true)
	if iocType != "" {
		query = query.Where("type = ?", iocType)
	}
	if value != "" {
		query = query.Where("value = ?", value)
	}

	var indicators []models.Indicator
	err := query.Order("created_at DESC").Limit(limit).Find(&indicators).Error
	return indicators, err
}

// CreateIndicator - Thêm IOC mới
func (s *IndicatorService) CreateIndicator(indicator *models.Indicator) error {
	indicator.ID = uuid.New().String()
	if indicator.RiskScore == 0 {
		indicator.RiskScore = 50
	}
	return s.db.Create(indicator).Error
}

func (s *IndicatorService) UpdateIndicator(id string, indicator *models.Indicator) error {
	if indicator == nil {
		return nil
	}
	if indicator.ID == "" {
		indicator.ID = id
	}
	result := s.db.Model(&models.Indicator{}).Where("id = ?", id).Updates(map[string]interface{}{
		"type":               indicator.Type,
		"value":              indicator.Value,
		"category":           indicator.Category,
		"source":             indicator.Source,
		"mitre_tactic":       indicator.MITRETactic,
		"mitre_technique_id": indicator.MITRETechniqueID,
		"risk_score":         indicator.RiskScore,
		"description":        indicator.Description,
		"is_active":          indicator.IsActive,
	})
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return gorm.ErrRecordNotFound
	}
	return nil
}

func (s *IndicatorService) DeleteIndicator(id string) error {
	result := s.db.Where("id = ?", id).Delete(&models.Indicator{})
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return gorm.ErrRecordNotFound
	}
	return nil
}

// SearchIndicatorValue - Tra cứu nhanh xem 1 giá trị (IP/Hash) có phải là IOC không
func (s *IndicatorService) SearchIndicatorValue(value string) (*models.Indicator, bool) {
	var indicator models.Indicator
	err := s.db.Where("value = ? AND is_active = ?", value, true).First(&indicator).Error
	if err != nil {
		return nil, false // Không tìm thấy hoặc đã lỗi/hết hạn
	}

	// Kiểm tra xem đã hết hạn chưa (nếu có set expires_at)
	if indicator.ExpiresAt != nil && time.Now().After(*indicator.ExpiresAt) {
		// Tự động disable nếu hết hạn
		s.db.Model(&indicator).Update("is_active", false)
		return nil, false
	}

	return &indicator, true
}

File: soc-server-python/app/services/indicator_service.py
# ==============================================================================
# app/services/indicator_service.py - Indicator Service (IOCs)
# Tương đương: internal/services/indicator_service.go
# ==============================================================================

import uuid
import logging
from datetime import datetime
from typing import Optional, Tuple, List, Dict, Any
from sqlalchemy import select, func, desc, delete
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.indicator import Indicator

logger = logging.getLogger(__name__)


def default_indicators() -> List[dict]:
    return [
        {
            "id": str(uuid.uuid4()),
            "type": "ip-src",
            "value": "185.220.101.182",
            "category": "c2_server",
            "source": "osint",
            "mitre_tactic": "Command and Control",
            "risk_score": 88,
            "description": "Known malicious IP used in phishing and command-and-control activity.",
            "is_active": True,
        },
        {
            "id": str(uuid.uuid4()),
            "type": "domain",
            "value": "login-secure-update.com",
            "category": "phishing",
            "source": "analyst",
            "mitre_tactic": "Phishing",
            "risk_score": 91,
            "description": "Phishing domain used in credential harvesting campaigns.",
            "is_active": True,
        },
        {
            "id": str(uuid.uuid4()),
            "type": "sha256",
            "value": "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08",
            "category": "malware",
            "source": "internal",
            "mitre_tactic": "Execution",
            "risk_score": 94,
            "description": "Known malicious payload hash observed in test environment.",
            "is_active": True,
        },
    ]


class IndicatorService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def get_all_indicators(
        self,
        page: int = 1,
        page_size: int = 50,
        ioc_type: str = "",
    ) -> Tuple[List[Indicator], int]:
        """Lấy danh sách IOCs với phân trang và filter"""
        stmt = select(Indicator)
        count_stmt = select(func.count(Indicator.id))

        if ioc_type:
            stmt = stmt.where(Indicator.type == ioc_type)
            count_stmt = count_stmt.where(Indicator.type == ioc_type)

        total_res = await self.db.execute(count_stmt)
        total = total_res.scalar() or 0

        offset = (page - 1) * page_size
        stmt = stmt.order_by(desc(Indicator.created_at)).offset(offset).limit(page_size)

        result = await self.db.execute(stmt)
        indicators = list(result.scalars().all())
        return indicators, total

    async def search_indicators_compat(
        self,
        ioc_type: str = "",
        value: str = "",
        limit: int = 50,
    ) -> List[Indicator]:
        """Tìm IOCs tương thích với node MISP của n8n"""
        limit = max(1, min(limit, 1000))
        stmt = select(Indicator).where(Indicator.is_active.is_(True))

        if ioc_type:
            stmt = stmt.where(Indicator.type == ioc_type)
        if value:
            stmt = stmt.where(Indicator.value == value)

        stmt = stmt.order_by(desc(Indicator.created_at)).limit(limit)
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def create_indicator(self, data: Dict[str, Any]) -> Indicator:
        """Thêm IOC mới"""
        indicator = Indicator(
            id=data.get("id") or str(uuid.uuid4()),
            type=data["type"],
            value=data["value"],
            category=data.get("category", "malware"),
            source=data.get("source", "analyst"),
            mitre_tactic=data.get("mitre_tactic"),
            mitre_technique_id=data.get("mitre_technique_id"),
            risk_score=data.get("risk_score", 50),
            description=data.get("description", ""),
            is_active=data.get("is_active", True),
            expires_at=data.get("expires_at"),
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        self.db.add(indicator)
        await self.db.commit()
        await self.db.refresh(indicator)
        return indicator

    async def update_indicator(self, indicator_id: str, data: Dict[str, Any]) -> Optional[Indicator]:
        """Cập nhật IOC"""
        stmt = select(Indicator).where(Indicator.id == indicator_id)
        res = await self.db.execute(stmt)
        indicator = res.scalar_one_or_none()
        if not indicator:
            return None

        for field in [
            "type", "value", "category", "source", "mitre_tactic",
            "mitre_technique_id", "risk_score", "description", "is_active"
        ]:
            if field in data and data[field] is not None:
                setattr(indicator, field, data[field])

        indicator.updated_at = datetime.utcnow()
        await self.db.commit()
        await self.db.refresh(indicator)
        return indicator

    async def delete_indicator(self, indicator_id: str) -> bool:
        """Xóa IOC"""
        stmt = delete(Indicator).where(Indicator.id == indicator_id)
        res = await self.db.execute(stmt)
        await self.db.commit()
        return (res.rowcount or 0) > 0

    async def search_indicator_value(self, value: str) -> Optional[Indicator]:
        """Tra cứu nhanh xem 1 giá trị (IP/Hash) có phải là IOC không"""
        stmt = select(Indicator).where(
            Indicator.value == value,
            Indicator.is_active.is_(True)
        )
        res = await self.db.execute(stmt)
        indicator = res.scalar_one_or_none()

        if not indicator:
            return None

        # Kiểm tra hết hạn nếu có expires_at
        if indicator.expires_at and datetime.utcnow() > indicator.expires_at:
            indicator.is_active = False
            await self.db.commit()
            return None

        return indicator
